book/src/proposals/tower-lockout-slashing.md

```diff
@@ -10,9 +10,10 @@ slashed by some percentage.
 ## Votes
 
 Validators submit votes which contain a list of slots.  To be
-accepted, the vote must contain the list of slots currently in the
-VoteState program, along with new votes.  The slots must match the
-VoteState exactly.
+accepted, the vote must contain the list of N slots currently in
+the VoteState program, along with new votes.  The first `N - new
+votes`, must match the VoteState exactly.  N should be large enough
```

Comment:
The N slots will also include the lockout right?  For instance,

vote 0, existing slots: {0, 1, 2}, new slot: {9}
vote 1: existing slots: {0, 1}, new slot: {5}

vote 0 implies that the lockout on slot 2 is at least 8, which would mean vote 1 is a violation.

To detect this condition, whatever sentry is tracking these votes have to track: 
1) The greatest lockout for each vote on a potential tower, basically building the most conservative view of a tower for each validator.
2) Make sure validators don't lie in their votes about the amount of lockout for a slot. For instance if the sentry sees for some validator A:

vote 0: 0, 1, 2, 

then they know that 0 has a lockout of at least 8 so they update for validator A, `lockout(0)` to 8 if `lockout(0) < 8` before

Reply:
@carllin 

` {0, 1, 2}, new slot: {9}`, 2 has a lockout of 2, 0 has a lockout of 8.
 `{0, 1}, new slot: {5}`

We don't need to show the "new slot".  You can look at just the lockouts and compute the slashing.  If the legth is greater that the epoch, then the min lockout on the first vote is the epoch.

Reply:
so, `0,1,2,9` would actually be `9`, because all the other votes have expired, and `0,5`.  You can compute the lockout for 0 from the gap between 0 and 5.

Reply:
hmmm 0, 1, 2, 9, the other votes wouldn't have expired if there had been previous votes on top of 2 that 9 popped off right? so if my previous stack had been `0, 1, 2, 3, 4`, and now I could have `0, 1, 2, 9`

Reply:
@carllin yea, thats correct.  

so "0,1,2,9" vs "0,1,2,3,4" shouldn't be slashed, since if you run locktower of the new votes from the first set on the second set, 3,4 would pop.  "0,1,2,5" wouldn't cause 3,4 to pop, so it must be a fork.

Reply:
hmmm, does the lockout always have to be the next power of 2? What happens in these cases:

Case 1:

`0, 1, 2`
`0, 8`
`0, 8, 9` 

The lockout on 0 is 8, but the distance to the top of the tower is 9. (Currently we don't pop from the back of the tower, so 0 wouldn't get popped by the vote for 9).

Case 2:
`0,1,2,3` 
`0,1,7`   
Distance to the top of the tower is 7, next power of two is 8, but the lockout on 0 is 16